smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java

```diff
@@ -0,0 +1,364 @@
+import org.jivesoftware.smack.XMPPException;
+import org.jivesoftware.smack.filter.AndFilter;
+import org.jivesoftware.smack.filter.StanzaFilter;
+import org.jivesoftware.smack.filter.StanzaTypeFilter;
+import org.jivesoftware.smack.packet.Message;
+import org.jivesoftware.smack.packet.MessageBuilder;
+import org.jivesoftware.smack.packet.Stanza;
+import org.jivesoftware.smack.packet.StanzaBuilder;
+import org.jivesoftware.smack.packet.XmlElement;
+
+import org.jivesoftware.smackx.disco.ServiceDiscoveryManager;
+import org.jivesoftware.smackx.disco.packet.DiscoverInfo;
+import org.jivesoftware.smackx.reactions.element.Reaction;
+import org.jivesoftware.smackx.reactions.element.ReactionsElement;
+import org.jivesoftware.smackx.reactions.filter.ReactionsFilter;
+import org.jivesoftware.smackx.xdata.FormField;
+import org.jivesoftware.smackx.xdata.TextSingleFormField;
+import org.jivesoftware.smackx.xdata.form.Form;
+import org.jivesoftware.smackx.xdata.packet.DataForm;
+
+import org.jxmpp.jid.BareJid;
+import org.jxmpp.jid.EntityBareJid;
+
+/**
+ * Manages reactions in the XMPP protocol. This class allows adding, removing, and listening for reactions
+ * on messages, as well as managing restrictions on the number of reactions per user and allowed emojis.
+ * It also allows propagating these restrictions to other clients via XMPP service discovery.
+ *
+ * This class is based on the XEP-0444 extension protocol for reactions.
+ *
+ * @author Ismael Nunes Campos
+ *
+ * @see <a href="https://xmpp.org/extensions/xep-0444.html">XEP-0444 Message Reactions</a>
+ * @see ReactionsElement
+ * @see Reaction
+ */
+public final class ReactionsManager extends Manager {
+
+    private static final Map<XMPPConnection, ReactionsManager> INSTANCES = new WeakHashMap<>();
+
+    static {
+        XMPPConnectionRegistry.addConnectionCreationListener(new ConnectionCreationListener() {
+            @Override
+            public void connectionCreated(XMPPConnection connection) {
+                getInstanceFor(connection);
+            }
+        });
+    }
+
+    private static final String REACTIONS_RESTRICTIONS_NAMESPACE = "urn:xmpp:reactions:0:restrictions";
+    private final Set<ReactionsListener> listeners = new CopyOnWriteArraySet<>();
+    private final AsyncButOrdered<BareJid> asyncButOrdered = new AsyncButOrdered<>();
+    private final StanzaFilter reactionsElementFilter = new AndFilter(StanzaTypeFilter.MESSAGE, ReactionsFilter.INSTANCE);
+
+    /**
+     * Constructs an instance of the reactions manager and add ReactionsElement to disco features.
+     *
+     * @param connection The XMPP connection used by the manager.
+     */
+    public ReactionsManager(XMPPConnection connection) {
+        super(connection);
+        connection.addAsyncStanzaListener(this::reactionsElementListener, reactionsElementFilter);
+        ServiceDiscoveryManager sdm = ServiceDiscoveryManager.getInstanceFor(connection);
+        sdm.addFeature(ReactionsElement.NAMESPACE);
```

Comment:
I don't think it is sensible to announce the feature right away. What if the Smack user did not setup a listener or isn't even aware of reactions? Then Smack would still announce the feature signalling to others that I can understand reactions.
A typical Smack idiom would be to only start announcing the feature once a reaction listener has been setup.